File: Bridge/Bridge.go
package main

import "fmt"

type DrawAPI interface {
	drawCircle(radius,x,y int)
}


type RedCircle struct {

}

func(r *RedCircle)drawCircle(radius,x,y int){
	fmt.Println("Drawing Circle[ color: red, radius: ",radius, ", x: ",x,", y:",y,"]\r\n")
}

type GreenCircle struct {

}

func (g *GreenCircle)drawCircle(radius,x,y int){
	fmt.Println("Drawing Circle[ color: green, radius: ",radius, ", x: ",x,", y:",y,"]\r\n")
}

type InterShape interface {
	draw()
}


type Shape struct {
	drawAPI DrawAPI
}

func(s *Shape)init(drawAPI DrawAPI){
	s.drawAPI=drawAPI
}


type Circle struct {
	Shape
	x int
	y int
	radius int
}

func(c *Circle)init(x,y,radius int,drawAPI DrawAPI){
	c.Shape.init(drawAPI)
	c.x=x
	c.y=y
	c.radius=radius
}

func(c *Circle)draw(){
	c.drawAPI.drawCircle(c.radius,c.x,c.y)
}


func main(){
	redCircle:=new(Circle)
	redCircle.init(100,100,10,new(RedCircle))
	greenCircle:=new(Circle)
	greenCircle.init(150,150,15,new(GreenCircle))
	redCircle.draw()
	greenCircle.draw()
}

File: Bridge/Bridge.js
/**
 * Created by jamesyan on 2017/3/19.
 */

/**
 interface DrawApi
 {
     public function drawCircle($radius, $x, $y);
 }

 type InterShape interface {
	draw()
}

 **/

var RedCircle=function(){};
RedCircle.prototype.drawCircle=function(radius,x,y){
    console.log("Drawing Circle[ color: red, radius: ",radius, ", x: ",x,", y:",y,"]\r\n");
}

var GreenCircle=function(){};

GreenCircle.prototype.drawCircle=function(radius,x,y){
    console.log("Drawing Circle[ color: green, radius: ",radius, ", x: ",x,", y:",y,"]\r\n");
}


var Shape=function(drawAPI){
    this.drawAPI=drawAPI;
}

var Circle=function(x,y,radius,drawAPI){
    Shape.call(this,drawAPI);
    this.x=x;
    this.y=y;
    this.radius=radius;
}

Circle.prototype.draw=function(){
    this.drawAPI.drawCircle(this.radius,this.x,this.y)
}

